providers/powerdns/powerdnsProvider.go

```diff
@@ -0,0 +1,269 @@
+
+// GetDomainCorrections returns a list of corrections to update a domain.
+func (api *PowerDNS) GetDomainCorrections(dc *models.DomainConfig) ([]*models.Correction, error) {
+	var corrections []*models.Correction
+
+	// record corrections
+	curRecords, err := api.GetZoneRecords(dc.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	// post-process records
+	dc.Punycode()
+	models.PostProcessRecords(curRecords)
+
+	// create record diff by group
+	keysToUpdate := (diff.New(dc)).ChangedGroups(curRecords)
+	desiredRecords := dc.Records.GroupedByKey()
+
+	// create corrections by group
+	for label, msgs := range keysToUpdate {
+		labelName := ensureDotSuffix(label.NameFQDN)
+		labelType := label.Type
+
+		if _, ok := desiredRecords[label]; !ok {
+			// nothing found, must be a delete
+			corrections = append(corrections, &models.Correction{
+				Msg: concatMessage(msgs),
+				F: func() error {
+					return api.client.Zones().RemoveRecordSetFromZone(context.Background(), api.ServerName, dc.Name, labelName, labelType)
+				},
+			})
+		} else {
+			ttl := desiredRecords[label][0].TTL
+			records := []zones.Record{}
+			for _, recordContent := range desiredRecords[label] {
+				records = append(records, zones.Record{
+					Content: recordContent.GetTargetCombined(),
+				})
+			}
+			corrections = append(corrections, &models.Correction{
+				Msg: concatMessage(msgs),
+				F: func() error {
+					return api.client.Zones().AddRecordSetToZone(context.Background(), api.ServerName, dc.Name, zones.ResourceRecordSet{
+						Name:    labelName,
+						Type:    labelType,
+						TTL:     int(ttl),
+						Records: records,
+					})
+				},
+			})
+		}
+	}
+
+	// DNSSec corrections
+	// TODO: Implementing of on-demand DNSSec creation in library
+
+	return corrections, nil
+}
+
+// EnsureDomainExists adds a domain to the DNS service if it does not exist
+func (api *PowerDNS) EnsureDomainExists(domain string) error {
+	if zone, _ := api.client.Zones().GetZone(context.Background(), api.ServerName, domain); zone.ID != "" {
+		return nil
+	}
+
+	_, err := api.client.Zones().CreateZone(context.Background(), api.ServerName, zones.Zone{
+		Name:        domain,
+		Type:        zones.ZoneTypeZone,
+		DNSSec:      api.DNSSecOnCreate,
+		Nameservers: api.DefaultNS,
+	})
+	return err
+}
+
+// toRecordConfig converts a PowerDNS DNSRecord to a RecordConfig. #rtype_variations
+func toRecordConfig(domain string, r zones.Record, ttl int, name string, rtype string) (*models.RecordConfig, error) {
+	// trimming trailing dot and domain from name
+	if strings.HasSuffix(name, domain+".") {
+		name = strings.TrimSuffix(name, domain+".")
+	}
+	if strings.HasSuffix(name, ".") {
+		name = strings.TrimSuffix(name, ".")
+	}
+
+	rc := &models.RecordConfig{
+		TTL:      uint32(ttl),
+		Original: r,
+		Type:     rtype,
+	}
+	rc.SetLabel(name, domain)
+
+	content := r.Content
+	switch rtype {
+	case "CNAME", "NS":
+		return rc, rc.SetTarget(dnsutil.AddOrigin(ensureDotSuffix(content), domain))
```

Comment:
I'm not sure I understand the logic here:

* The code adds a dot, then calls AddOrigin, which is a no-op if the string ends in a dot.

    rc.SetTarget(ensureDotSuffix(content))

Now that it is simplified, we might be able to go even more simple.

Does the API return content with a trailing dot sometimes but not others?  Old dnscontrol code used to be paranoid and handle everything both ways.  Eventually we found the complexity wasn't worth it.  APIs would never change such a fundamental thing. 

Unless you have proof the API sometimes returns with or without a trailing dot, this code could just be:

		return rc, rc.SetTarget(content + ".")

That said... test this.  I could always be wrong :-)

Reply:
I've tested it with a other version of PDNS, both returned it with a dot in the end. So i removed it.